golem-component-service-base/src/repo/component.rs

```diff
@@ -0,0 +1,467 @@
+    fn from(value: ComponentRecord) -> Self {
+        VersionedComponentId {
+            component_id: ComponentId(value.component_id),
+            version: value.version as u64,
+        }
+    }
+}
+
+impl ComponentRecord {
+    pub fn new<Namespace: Display>(
+        namespace: Namespace,
+        component: Component,
+    ) -> Result<Self, String> {
+        let metadata = record_metadata_serde::serialize(&component.metadata)?;
+        Ok(Self {
+            namespace: namespace.to_string(),
+            component_id: component.versioned_component_id.component_id.0,
+            name: component.component_name.0,
+            size: component.component_size as i32,
+            version: component.versioned_component_id.version as i64,
+            user_component: component.versioned_component_id.slug(),
+            protected_component: component.protected_component_id.slug(),
+            protector_version: None,
+            metadata: metadata.into(),
+        })
+    }
+}
+
+#[async_trait]
+pub trait ComponentRepo {
+    async fn upsert(&self, component: &ComponentRecord) -> Result<(), RepoError>;
+
+    async fn get(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_all(&self, namespace: &str) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_latest_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Option<ComponentRecord>, RepoError>;
+
+    async fn get_by_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+        version: u64,
+    ) -> Result<Option<ComponentRecord>, RepoError>;
+
+    async fn get_by_name(
+        &self,
+        namespace: &str,
+        name: &str,
+    ) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_ids_by_name(&self, namespace: &str, name: &str) -> Result<Vec<Uuid>, RepoError>;
+
+    async fn get_namespaces(&self, component_id: &Uuid) -> Result<Vec<(String, i64)>, RepoError>;
+
+    async fn delete(&self, namespace: &str, component_id: &Uuid) -> Result<(), RepoError>;
+}
+
+pub struct DbComponentRepo<DB: Database> {
+    db_pool: Arc<Pool<DB>>,
+}
+
+impl<DB: Database> DbComponentRepo<DB> {
+    pub fn new(db_pool: Arc<Pool<DB>>) -> Self {
+        Self { db_pool }
+    }
+}
+
+#[async_trait]
+impl ComponentRepo for DbComponentRepo<sqlx::Sqlite> {
+    async fn upsert(&self, component: &ComponentRecord) -> Result<(), RepoError> {
+        let mut transaction = self.db_pool.begin().await?;
+
+        let result = sqlx::query(
+            "SELECT count(component_id) as count FROM components WHERE namespace != $1 AND component_id = $2",
+        )
+        .bind(component.namespace.clone())
+        .bind(component.component_id)
+        .fetch_one(&mut *transaction)
+        .await?;
+
+        let count: i64 = result.get("count");
+
+        if count > 0 {
```

Comment:
There is a race here - you can create 2 different versions in different namespaces if you have 2 `upsert` running at the same same. The probability is low, but the race is there. The only solution without race is to normalise the schema.